components/browser/session/src/main/java/mozilla/components/browser/session/storage/SessionStorage.kt

```diff
@@ -4,28 +4,39 @@
 
 package mozilla.components.browser.session.storage
 
+import mozilla.components.browser.session.Session
 import mozilla.components.browser.session.SessionManager
+import mozilla.components.concept.engine.Engine
+import mozilla.components.concept.engine.EngineSession
+
+data class SessionWithState(
+    val session: Session,
+    val selected: Boolean = false,
+    val engineSession: EngineSession? = null
+)
+typealias SessionsSnapshot = List<SessionWithState>
 
 /**
  * Storage component for browser and engine sessions.
  */
 interface SessionStorage {
 
     /**
-     * Persists the session state of the provided [SessionManager].
+     * Persists the provided [SessionsSnapshot] for a given [Engine].
      *
-     * @param sessionManager the session manager to persist from.
-     * @return true if the state was persisted, otherwise false.
+     * @param engine the session in which context to persist a snapshot.
+     * @param sessions the snapshot of sessions which are to be persisted.
+     * @return true if the snapshot was persisted, otherwise false.
      */
-    fun persist(sessionManager: SessionManager): Boolean
+    fun persist(engine: Engine, sessions: SessionsSnapshot): Boolean
 
     /**
-     * Restores the session state by reading from the latest persisted version.
+     * Returns the latest persisted sessions snapshot that may be used to restore sessions.
      *
-     * @param sessionManager the session manager to restore into.
-     * @return true if the state was restored, otherwise false.
+     * @param engine the engine for which to read the snapshot.
+     * @return snapshot of sessions to restore, or null if they couldn't be read.
      */
-    fun restore(sessionManager: SessionManager): Boolean
+    fun restore(engine: Engine): SessionsSnapshot?
```

Comment:
Now that `SessionManager` has a `restore` method, I prefer this one to be `read`, but that's a nit. It would give us:
```Kotlin
sessionStorage.read(engine)?.let { restore(it) }
```
vs.
```Kotlin
sessionStorage.restore(engine)?.let { restore(it) }
```
Whatever you prefer though. Just a comment.

Reply:
Yeah, the restore & restore is a pretty weird combo - and that name isn’t factually correct, either. I’d prefer read as well.